log/log_test.go

```diff
@@ -0,0 +1,301 @@
+	_, err = ginWriter.Write([]byte("[WARNING] message"))
+	require.NoError(t, err)
+
+	_, err = ginWriter.Write([]byte("[ERROR] message"))
+	require.NoError(t, err)
+
+	require.NoError(t, writer.Sync())
+
+	expected := "\x1b[32mINFO\x1b[0m\ttest\tnormal message\n\x1b[32mINFO\x1b[0m\ttest\t\x1b[33m[WARNING]\x1b[0m message\n\x1b[32mINFO\x1b[0m\ttest\t\x1b[31m[ERROR]\x1b[0m message\n"
+
+	assert.Equal(t, expected, string(output))
+}
+
+func Test_Color(t *testing.T) {
+	var enc TestArrayEncoder
+
+	color.NoColor = false
+
+	ColorCapitalLevelEncoder(zap.InfoLevel, &enc)
+	require.Len(t, enc.Elems, 1)
+
+	expected := []byte("\x1b[32mINFO\x1b[0m")
+	actual := []byte(enc.Elems[0].(string))
+	assert.Equal(t, expected, actual)
+}
+
+// Adapted from zap internal implementation.
+type TestArrayEncoder struct {
+	Elems []interface{}
+}
+
+func (s *TestArrayEncoder) AppendReflected(v interface{}) error {
+	s.Elems = append(s.Elems, v)
+	return nil
+}
+
+func (s *TestArrayEncoder) AppendArray(v zapcore.ArrayMarshaler) error   { return nil }
+func (s *TestArrayEncoder) AppendObject(v zapcore.ObjectMarshaler) error { return nil }
+
+func (s *TestArrayEncoder) AppendBool(v bool)              { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendByteString(v []byte)      { s.Elems = append(s.Elems, string(v)) }
+func (s *TestArrayEncoder) AppendComplex128(v complex128)  { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendComplex64(v complex64)    { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendDuration(v time.Duration) { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendFloat64(v float64)        { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendFloat32(v float32)        { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendInt(v int)                { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendInt64(v int64)            { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendInt32(v int32)            { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendInt16(v int16)            { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendInt8(v int8)              { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendString(v string)          { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendTime(v time.Time)         { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendUint(v uint)              { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendUint64(v uint64)          { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendUint32(v uint32)          { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendUint16(v uint16)          { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendUint8(v uint8)            { s.Elems = append(s.Elems, v) }
+func (s *TestArrayEncoder) AppendUintptr(v uintptr)        { s.Elems = append(s.Elems, v) }
```

Comment:
To implement the `zapcore.PrimitiveArrayEncoder` interface, which is what ` ColorCapitalLevelEncoder` expects (which it does, because it itself needs to implement `zap.LevelEncoder`).

Reply:
🤯